kernel/virtio/virtio_ring.c

```diff
@@ -0,0 +1,145 @@
+/* Copyright (c) 2015, IBM
+ * Author(s): Dan Williams <djwillia@us.ibm.com>
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "kernel.h"
+#include "virtio_ring.h"
+#include "virtio_pci.h"
+
+/*
+ * There is no official max queue size. But we've seen 4096, so let's use the
+ * double of that.
+ */
+#define VIRTQ_NET_MAX_QUEUE_SIZE 8192
+
+/* WARNING: called in interrupt context */
+void virtq_handle_interrupt(struct virtq *vq)
+{
+    volatile struct virtq_used_elem *e;
+
+    for (;;) {
+        uint16_t desc_idx;
+        struct io_buffer *head_buf;
+
+        if ((vq->used->idx % vq->num) == vq->last_used)
```

Comment:
Is last_used free-running? Code in drivers uses (% num) when touching it, doesn't it need the same here? (Also see general comment coming up)